packages/contracts/src/interest/MultiTokenVault.sol

```diff
@@ -1,232 +1,280 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.20;
+pragma solidity ^0.8.23;
 
 import { IMultiTokenVault } from "@credbull/interest/IMultiTokenVault.sol";
-import { ERC1155 } from "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import { ERC1155Supply, ERC1155 } from "@openzeppelin/contracts/token/ERC1155/extensions/ERC1155Supply.sol";
 import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
-import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
-import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { IERC165 } from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
 
 /**
  * @title MultiTokenVault
- * @dev A vault that uses SimpleInterest and Discounting to calculate shares per asset.
- *      The vault manages deposits and redemptions based on elapsed time periods and applies simple interest calculations.
+ * @dev A vault that uses deposit-period-specific ERC1155 tokens to represent deposits.
+ *      This contract manages deposits and redemptions using ERC1155 tokens. It tracks the number
+ *      of time periods that have elapsed and allows users to deposit and redeem assets based on these periods.
+ *      Designed to be secure and production-ready for Hacken audit.
  */
-abstract contract MultiTokenVault is IMultiTokenVault, ERC1155 {
-    using Math for uint256;
+abstract contract MultiTokenVault is ERC1155Supply, IMultiTokenVault, ReentrancyGuard, Ownable {
```

Comment:
a few points related to Access Control:

1. in OpenZeppelin contracts Access Control and Ownable are separated from their Token contracts such as ERC1555 and required in child implementations.  suggest we follow that.
2. I prefer AccessControl over Ownable as it gives us more fine-grained control.  please refer to the CBL token for example where we have an Owner and a Minter.